streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -337,8 +337,38 @@ class ZSink[-R, -InErr, -In, +OutErr, +L, +Z](val channel: ZChannel[R, InErr, Ch
    */
   def untilOutputM[R1 <: R, OutErr1 >: OutErr](
     f: Z => ZIO[R1, OutErr1, Boolean]
-  )(implicit ev: L <:< In): ZSink[R1, InErr, In, OutErr1, L, Option[Z]] =
-    ???
+  )(implicit ev: L <:< In): ZSink[R1, InErr, In, OutErr1, L, Option[Z]] = {
+
+    def processChunk(in: Chunk[In]): ZChannel[R1, InErr, Chunk[In], Any, OutErr1, Chunk[L], Option[Z]] =
+      (ZChannel.write(in) >>> self.channel).doneCollect.map { case (chunks, z) =>
```

Comment:
If I understand this correctly, this is subtly wrong because it assumes that the `self` sink only needs to read one chunk to complete.

It's worth testing this with a sink that needs two chunks to complete; e.g. ZSink.take(4) on a stream with two chunks of two elements.